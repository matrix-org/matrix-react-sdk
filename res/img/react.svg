<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
