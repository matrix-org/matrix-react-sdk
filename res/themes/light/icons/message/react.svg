<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>