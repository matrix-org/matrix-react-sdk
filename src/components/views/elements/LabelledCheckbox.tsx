/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import classnames from "classnames";

import StyledCheckbox from "./StyledCheckbox";

interface IProps {
    // The value for the checkbox
    value: boolean;
    // The translated label for the checkbox
    label: string;
    // Optional translated string to show below the checkbox
    byline?: string;
    // Whether or not to disable the checkbox
    disabled?: boolean;
    // The function to call when the value changes
    onChange(checked: boolean): void;
    // Optional additional CSS class to apply to the label
    className?: string;
}

const LabelledCheckbox: React.FC<IProps> = ({ value, label, byline, disabled, onChange, className }) => {
    return (
        <label className={classnames("mx_LabelledCheckbox", className)}>
            <StyledCheckbox disabled={disabled} checked={value} onChange={(e) => onChange(e.target.checked)} />
            <div className="mx_LabelledCheckbox_labels">
                <span className="mx_LabelledCheckbox_label">{label}</span>
                {byline ? <span className="mx_LabelledCheckbox_byline">{byline}</span> : null}
            </div>
        </label>
    );
};

export default LabelledCheckbox;
