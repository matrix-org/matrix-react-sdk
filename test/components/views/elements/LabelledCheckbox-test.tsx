/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";

import LabelledCheckbox from "../../../../src/components/views/elements/LabelledCheckbox";

describe("<LabelledCheckbox />", () => {
    type CompProps = React.ComponentProps<typeof LabelledCheckbox>;
    const getComponent = (props: CompProps) => <LabelledCheckbox {...props} />;
    const getCheckbox = (): HTMLInputElement => screen.getByRole("checkbox");

    it.each([undefined, "this is a byline"])("should render with byline of %p", (byline) => {
        const props: CompProps = {
            label: "Hello world",
            value: true,
            byline: byline,
            onChange: jest.fn(),
        };
        const renderResult = render(getComponent(props));
        expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it("should support unchecked by default", () => {
        const props: CompProps = {
            label: "Hello world",
            value: false,
            onChange: jest.fn(),
        };
        render(getComponent(props));
        expect(getCheckbox()).not.toBeChecked();
    });

    it("should be possible to disable the checkbox", () => {
        const props: CompProps = {
            label: "Hello world",
            value: false,
            disabled: true,
            onChange: jest.fn(),
        };
        render(getComponent(props));
        expect(getCheckbox()).toBeDisabled();
    });

    it("should emit onChange calls", () => {
        const props: CompProps = {
            label: "Hello world",
            value: false,
            onChange: jest.fn(),
        };
        render(getComponent(props));

        expect(props.onChange).not.toHaveBeenCalled();
        fireEvent.click(getCheckbox());
        expect(props.onChange).toHaveBeenCalledWith(true);
    });

    it("should react to value and disabled prop changes", () => {
        const props: CompProps = {
            label: "Hello world",
            value: false,
            onChange: jest.fn(),
        };
        const { rerender } = render(getComponent(props));

        let checkbox = getCheckbox();
        expect(checkbox).not.toBeChecked();
        expect(checkbox).not.toBeDisabled();

        props.disabled = true;
        props.value = true;
        rerender(getComponent(props));

        checkbox = getCheckbox();
        expect(checkbox).toBeChecked();
        expect(checkbox).toBeDisabled();
    });

    it("should render with a custom class name", () => {
        const className = "some class name";
        const props: CompProps = {
            label: "Hello world",
            value: false,
            onChange: jest.fn(),
            className,
        };
        const { container } = render(getComponent(props));
        expect(container.firstElementChild?.className).toContain(className);
    });
});
