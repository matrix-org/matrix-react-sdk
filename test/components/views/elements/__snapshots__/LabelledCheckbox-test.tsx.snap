// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelledCheckbox /> should render with byline of "this is a byline" 1`] = `
<LabelledCheckbox
  byline="this is a byline"
  label="Hello world"
  onChange={[MockFunction]}
  value={true}
>
  <label
    className="mx_LabelledCheckbox"
  >
    <StyledCheckbox
      checked={true}
      className=""
      onChange={[Function]}
    >
      <span
        className="mx_Checkbox mx_Checkbox_hasKind mx_Checkbox_kind_solid"
      >
        <input
          checked={true}
          id="checkbox_abdefghi"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="checkbox_abdefghi"
        >
          <div
            className="mx_Checkbox_background"
          >
            <div
              className="mx_Checkbox_checkmark"
            />
          </div>
        </label>
      </span>
    </StyledCheckbox>
    <div
      className="mx_LabelledCheckbox_labels"
    >
      <span
        className="mx_LabelledCheckbox_label"
      >
        Hello world
      </span>
      <span
        className="mx_LabelledCheckbox_byline"
      >
        this is a byline
      </span>
    </div>
  </label>
</LabelledCheckbox>
`;

exports[`<LabelledCheckbox /> should render with byline of null 1`] = `
<LabelledCheckbox
  byline={null}
  label="Hello world"
  onChange={[MockFunction]}
  value={true}
>
  <label
    className="mx_LabelledCheckbox"
  >
    <StyledCheckbox
      checked={true}
      className=""
      onChange={[Function]}
    >
      <span
        className="mx_Checkbox mx_Checkbox_hasKind mx_Checkbox_kind_solid"
      >
        <input
          checked={true}
          id="checkbox_abdefghi"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="checkbox_abdefghi"
        >
          <div
            className="mx_Checkbox_background"
          >
            <div
              className="mx_Checkbox_checkmark"
            />
          </div>
        </label>
      </span>
    </StyledCheckbox>
    <div
      className="mx_LabelledCheckbox_labels"
    >
      <span
        className="mx_LabelledCheckbox_label"
      >
        Hello world
      </span>
    </div>
  </label>
</LabelledCheckbox>
`;
